Detect (new) Bloom/Ribbon Filter construction corruption

hx235 opened the pull request:
Note: rebase on and merge after https://github.com/facebook/rocksdb/pull/9349, https://github.com/facebook/rocksdb/pull/9345, (optional) https://github.com/facebook/rocksdb/pull/9393 
**Context:**
(Quoted from @pdillinger) Layers of information during new Bloom/Ribbon Filter construction in building block-based tables includes the following: 
a) set of keys to add to filter
b) set of hashes to add to filter (64-bit hash applied to each key)
c) set of Bloom indices to set in filter, with duplicates
d) set of Bloom indices to set in filter, deduplicated
e) final filter and its checksum

This PR aims to detect corruption (e.g, unexpected hardware/software corruption on data structures residing in the memory for a long time) from b) to e) and leave a) as future works for application level. 
- b)'s corruption is detected by verifying the xor checksum of the hash entries calculated as the entries accumulate before being added to the filter. (i.e, `XXPH3FilterBitsBuilder::MaybeVerifyHashEntriesChecksum()`)
- c) - e)'s corruption is detected by verifying the hash entries indeed exists in the constructed filter by re-querying these hash entries in the filter (i.e, `FilterBitsBuilder::MaybePostVerify()`) after computing the block checksum (except for PartitionFilter, which is done right after each `FilterBitsBuilder::Finish` for impl simplicity - see code comment for more). For this stage of detection, we assume hash entries are not corrupted after checking on b) since the time interval from b) to c) is relatively short IMO.

Option to enable this feature of detection is `BlockBasedTableOptions::detect_filter_construct_corruption` which is false by default.

**Summary:**
- Implemented new functions `XXPH3FilterBitsBuilder::MaybeVerifyHashEntriesChecksum()` and `FilterBitsBuilder::MaybePostVerify()`
- Ensured hash entries, final filter and banding and their [cache reservation ](#9073) are released properly despite corruption
   - See [Filter.construction.artifacts.release.point.pdf ](https://github.com/facebook/rocksdb/files/7923487/Design.Filter.construction.artifacts.release.point.pdf)for high-level design
   -  Bundled and refactored hash entries's related artifact in XXPH3FilterBitsBuilder into `HashEntriesInfo` for better control on lifetime of these artifact during `SwapEntires`, `ResetEntries`
- Ensured RocksDB block-based table builder calls `FilterBitsBuilder::MaybePostVerify()` after constructing the filter by `FilterBitsBuilder::Finish()`
- When encountering such filter construction corruption, stop writing the filter content to files and mark such a block-based table building non-ok by storing the corruption status in the builder.

**Test:**
- Added new unit test `DBFilterConstructionCorruptionTestWithParam.DetectCorruption`
- Included this new feature in `DBFilterConstructionReserveMemoryTestWithParam.ReserveMemory` as this feature heavily touch ReserveMemory's impl
   - For fallback case, I run `./filter_bench -impl=3 -detect_filter_construct_corruption=true -reserve_table_builder_memory=true -strict_capacity_limit=true  -quick -runs 10 | grep 'Build avg'` to make sure nothing break. 
- Added to `filter_bench`: increased filter construction time by **30%**, mostly by `MaybePostVerify()`
   -  FastLocalBloom
       - Before change: `./filter_bench -impl=2 -quick -runs 10 | grep 'Build avg'`: **28.86643s**
       - After change:
          -  `./filter_bench -impl=2 -detect_filter_construct_corruption=false -quick -runs 10 | grep 'Build avg'` (expect a tiny increase due to MaybePostVerify is always called regardless): **27.6644s (-4% perf improvement might be due to now we don't drop bloom hash entry in `AddAllEntries` along iteration but in bulk later, same with the bypassing-MaybePostVerify case below)**
          - `./filter_bench -impl=2 -detect_filter_construct_corruption=true -quick -runs 10 | grep 'Build avg'` (expect acceptable increase): **34.41159s (+20%)**
          - `./filter_bench -impl=2 -detect_filter_construct_corruption=true -quick -runs 10 | grep 'Build avg'` (by-passing MaybePostVerify, expect minor increase): **27.13431s (-6%)**
    -  Standard128Ribbon
       - Before change: `./filter_bench -impl=3 -quick -runs 10 | grep 'Build avg'`: **122.5384s**
       - After change:
          - `./filter_bench -impl=3 -detect_filter_construct_corruption=false -quick -runs 10 | grep 'Build avg'` (expect a tiny increase due to MaybePostVerify is always called regardless - verified by removing MaybePostVerify under this case and found only +-1ns difference): **124.3588s (+2%)**
          - `./filter_bench -impl=3 -detect_filter_construct_corruption=true -quick -runs 10 | grep 'Build avg'`(expect acceptable increase): **159.4946s (+30%)**
          - `./filter_bench -impl=3 -detect_filter_construct_corruption=true -quick -runs 10 | grep 'Build avg'`(by-passing MaybePostVerify, expect minor increase) : **125.258s (+2%)**
- Added to `db_stress`: `make crash_test`, `./db_stress --detect_filter_construct_corruption=true`
- Manually smoke-tested: manually corrupted the filter construction in some db level tests with basic PUT and background flush. As expected, the error did get returned to users in subsequent PUT and Flush status.

hx235 commented on include/rocksdb/filter_policy.h:
```diff
@@ -65,6 +65,32 @@ class FilterBitsBuilder {
+  // for a non-null status pointer argument, it should point to
+  // Status::Corruption() when there is any corruption during filter
+  // construction or Status::OK() otherwise.
+  virtual Slice Finish(std::unique_ptr<const char[]>* buf,
```
Note: I decided to add a status output argument in `Finish` to communicate corruption status. Another solution is to rely on the length of filter content being 0, which might collide with the `hash_entries_.empty() == true` [case](https://github.com/facebook/rocksdb/blob/6.27.fb/table/block_based/filter_policy.cc#L508-L512) in Ribbon filter. It is also less understandable and less standard in my opinion.

Downside of the current approach is name collision of two Finish function could trigger `-Werror=overloaded-virtual`. We could change the function name to `FinishWithStatus` like `[GetBuilderWithContext](https://github.com/facebook/rocksdb/blob/6.27.fb/include/rocksdb/filter_policy.h#L203-L208)` to avoid this. But considering it might be an infrequent case to customize FilterBitsBuilder since it's almost expert work to do so, we might be able to also leave this consequence as it is.

hx235 commented on table/block_based/filter_policy.cc:
```diff
@@ -1107,14 +1242,54 @@ class AlwaysTrueFilter : public FilterBitsReader {
+    return s;
+  }
+
+  std::unique_ptr<FilterBitsReader> bits_reader(
+      filter_policy_->GetFilterBitsReader(filter_content));
+  auto sync_point_callback_arg =
+      std::make_pair(&bits_reader, filter_content.ToString());
+  TEST_SYNC_POINT_CALLBACK(
+      "XXPH3FilterBitsBuilder::PostVerify::PreVerification",
+      &sync_point_callback_arg);
+
+  if (bits_reader->IsAlwaysTrueFilter()) {
+    // BuiltinFilterPolicy::GetFilterBitsReader() returns AlwaysTrueFilter for
+    // compability of filter content not supported by FastLocalBloomBitsReader
+    // nor Standard128RibbonBitsReader. However, such filter content should not
+    // be passed into XXPH3FilterBitsBuilder::PostVerify unless it experienced
+    // an corruption
+    s = Status::Corruption("Corrupted filter content");
```
Note: I found it a lot cleaner to rely on `filter_policy_->GetFilterBitsReader(filter_content))` instead of manually creating FilterBitsReader according to the  filter_content, which duplicated some internal logic of `GetFilterBitsReader`. 

But there is one tricky point as described in comment in line 1274-1278 -`GetFilterBitsReader()` can `AlwaysTrueFilter()` for compatibility which defeats the purpose of detecting corruption in here. Therefore it asks for a way to differentiate `AlwaysTrueFilter` from other `FilterBitsReader` during run-time. 

To dd so and considering `RTTI` support can be unavailable according to `[..Replace dynamic_cast<> (except unit test) so people can choose to build with RTTI off. With make, release mode is by default built with -fno-rtti and debug mode is built without it. Users can override it by setting USE_RTTI=0 or 1.](https://github.com/facebook/rocksdb/blob/main/HISTORY.md#new-features-40)`, it seems that the only way left is to add some class info exposure method like `IsAlwaysTrueFilter()`. @mdcallag 's work in Customization in `FilterBitsBuilder` (if there is any) may also help in here.

pdillinger replied:
I haven't read all the details, but (in relation to #9412) it should be *safe* (though inefficient) to PostVerify an AlwaysTrueFilter, because all keys return `true` for MayMatch. I'm not sure it's worth carving out an exception for this obscure case.

hx235 replied:
(no need to reply)Update: discussed offline to clarify the potential cost of ignoring "corruption to AlwaysTrueFilter/LegacyBloomBitsReader" case and we agreed that the cost is acceptable with the benefit of simplicity w/o special case handling and customization #9412

hx235 commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -1547,7 +1547,13 @@ void BlockBasedTableBuilder::WriteFilterBlock(
           rep_->filter_builder->Finish(filter_block_handle, &s, &filter_data);
-      assert(s.ok() || s.IsIncomplete());
+
+      assert(s.ok() || s.IsIncomplete() || s.IsCorruption());
```
Note: 
Partitioned filters use `s.IsIncomplete()` to indicate more partitioned filter to proceed and we probably don't want filter construction somehow returns `Status::Incomplete` to indicate another thing. 

One way is to explicitly ask (by [documentation](https://github.com/facebook/rocksdb/pull/9342/files#diff-90d330d77a09b375bb5370051e4529ee67c905d7b72b2b599de722edab9a44f7R68-R71)) that filter construction can only return OK or Corruption.

Another way is to be very specific about the  `Status::Incomplete` used in partitioned filter and identify it using status message so that another `Status::Incomplete` returned from filter construction is allowed.

hx235 commented:
@pdillinger I figured it might be a good time to request some early review on a few API designs commented in the PR but Feel free to **delay** reviewing this after the break - requested now just in case I forget :)

Retrying on compaction on filter construction corruption seems to deserve another PR - hacked something here https://github.com/facebook/rocksdb/pull/9350 but needs to think more about it.

I also left two questions for clarification on follow-up in the task.

hx235 commented on table/table_test.cc:
```diff
@@ -2547,6 +2548,146 @@ void TableTest::IndexTest(BlockBasedTableOptions table_options) {
   c.ResetTableReader();
 }
 
+class BlockBasedTableFilterConstructCorruptionTest
```
TODO: If it's okay to separate re-trying compaction into another PR, then in this PR, it's worth verifying that adding a new type of error in table builder does not break existing high-level write API Compact, Put, Delete

hx235 replied:
Yes - it's okay to separate retrying compaction.

hx235 commented on table/block_based/full_filter_block.cc:
```diff
@@ -109,8 +109,11 @@ Slice FullFilterBlockBuilder::Finish(
+    Slice filter_content = filter_bits_builder_->Finish(
+        filter_data ? filter_data : &filter_data_, status);
+    if (status->ok()) {
+      *status = filter_bits_builder_->PostVerify(filter_content);
```
TODO:
Need to move the PostVerify after computing block checksum based on the task update

hx235 replied:
Done

hx235 commented on table/block_based/partitioned_filter_block.cc:
```diff
@@ -88,9 +88,15 @@ void PartitionedFilterBlockBuilder::MaybeCutAFilterBlock(
+  Slice filter =
+      filter_bits_builder_->Finish(&filter_data, &filter_construction_status);
+  if (filter_construction_status.ok()) {
+    filter_construction_status = filter_bits_builder_->PostVerify(filter);
```
TODO:
Need to move the PostVerify after computing block checksum based on the task update

hx235 replied:
Done

hx235 commented:
Update:
- rebase on https://github.com/facebook/rocksdb/pull/9349

hx235 commented on table/block_based/filter_policy.cc:
```diff
@@ -97,6 +109,24 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
   // For delegating between XXPH3FilterBitsBuilders
   void SwapEntriesWith(XXPH3FilterBitsBuilder* other) {
     std::swap(hash_entries_, other->hash_entries_);
+    if (cache_res_mgr_) {
+      std::swap(hash_entry_cache_res_bucket_handles_,
+                other->hash_entry_cache_res_bucket_handles_);
+    }
+    if (detect_filter_construct_corruption_) {
+      std::swap(hash_entries_checksum_, other->hash_entries_checksum_);
+    }
+    entries_swapped_ = true;
+  }
+
+  void ResetEntries() {
+    hash_entries_.clear();
+    if (cache_res_mgr_) {
+      hash_entry_cache_res_bucket_handles_.clear();
+    }
+    if (detect_filter_construct_corruption_) {
+      hash_entries_checksum_ = 0;
+    }
   }
```
Note: 
Seems like a good refactoring to have a Struct `HashEntriesInfo` grouping `hash_entries_ + hash_entry_cache_res_bucket_handles_ + hash_entries_checksum_` and do `HashEntriesInfo.Swap(..)`, `HashEntriesInfo.Reset()`instead.

hx235 replied:
Opened task

hx235 replied:
Done

hx235 commented:
Update:
Rebase

hx235 commented:
Update:
Just a rebase - will come back to 2 TODO left by me later.

hx235 commented:
Update: 
- Rebase 
- Call PostVerify after filter block checksum

hx235 commented:
Update: 
Still improving this PR

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -1111,14 +1242,54 @@ class AlwaysTrueFilter : public FilterBitsReader {
+Status XXPH3FilterBitsBuilder::PostVerify(const Slice& filter_content) {
+  Status s = Status::OK();
+
+  if (!detect_filter_construct_corruption_) {
```
I haven't read through all the details, but it would seem cleaner to me to make this decision (based on option) before calling PostVerify, so that custom FilterBitsBuilder doesn't have to be involved in that decision.

pdillinger replied:
Oh right I forgot about accumulating vs. purging keys, which means we want to know earlier whether we're going to verify.

hx235 replied:
(no need to reply) Discussed offline so we are still taking the MaybePostVerify approach with clarified comment as suggestion below

hx235 commented:
Updating this draft for a few fix @pdillinger - just in case you are reading it now

pdillinger commented on include/rocksdb/filter_policy.h:
```diff
@@ -65,6 +65,32 @@ class FilterBitsBuilder {
+  //
+  // The verification will only take effects
+  // when BlockBasedTableOptions::detect_filter_construct_corruption = true.
+  // Otherwise it is no-op and always returns Status::OK().
```
This should instead say something like "Implementations should normally consult FilterBuildingContext::table_options.detect_filter_construct_corruption to determine whether to perform verification or to skip by returning Status::OK(). The decision is left to the FilterBitsBuilder so that verification prerequisites before PostVerify can be skipped when not configured." Does my wording make sense?

hx235 commented:
Update:
- ci/circleci: build-macos  fails with an apparently irrelevant test 502 34051 34049   0  3:45AM ttys001    0:17.18 ./backupable_db_test --gtest_filter=BackupEngineTest.OpenBackupAsReadOnlyDB

Too long with no output (exceeded 10m0s): context deadline exceeded

hx235 commented on table/block_based/partitioned_filter_block.h:
```diff
@@ -41,6 +41,15 @@ class PartitionedFilterBlockBuilder : public FullFilterBlockBuilder {
+  // as part of PartitionFilterBlockBuilder::Finish
+  // to avoid implementation complexity of doing it elsewhere.
+  // Therefore we are skipping it in here.
+  virtual Status MaybePostVerifyFilter(
```
Note:
Relevant to my latest offline msg about PostVerify PartitionFilter after block checksum, I decided to PostVerify inside of PartitionFilter::Finish() instead of outside the function. I figured this deserves a note in the API, even though technically we don't have to override MaybePostVerifyFilter.

hx235 commented on table/block_based/partitioned_filter_block.cc:
```diff
@@ -88,9 +90,18 @@ void PartitionedFilterBlockBuilder::MaybeCutAFilterBlock(
+  }
   std::string& index_key = p_index_builder_->GetPartitionKey();
-  filters.push_back({index_key, filter, std::move(filter_data)});
+  filters.push_back({index_key, std::move(filter_data), filter});
+  if (!filter_construction_status.ok() &&
+      partitioned_filters_construction_status_.ok()) {
+    partitioned_filters_construction_status_ = filter_construction_status;
+  }
```
Note: I wish I could even not producing the rest of the partitioned filters once I encountered a filter construction error on some filter partition. However, the PartitionFilter Add and Finish logic isn't very easy to understand so I decided to go for a less aggressive way of responding to some partition construction error.

hx235 commented on db_stress_tool/db_stress_common.h:
```diff
@@ -150,6 +150,7 @@ DECLARE_bool(use_block_based_filter);
 DECLARE_int32(ribbon_starting_level);
 DECLARE_bool(partition_filters);
 DECLARE_bool(optimize_filters_for_memory);
+DECLARE_bool(detect_filter_construct_corruption);
```
Question: 
I feel like it does not hard to add this to stress test since we can test it with other random options. It is just that I don't know what to expect from stress test IF this feature did detect a filter construction corruption. For example, I manually injected some construction error and it gives me terminate and core dump when running `./db_stress --detect_filter_construct_corruption=true` - is it what we want from stress test?

hx235 replied:
Hmmm .. just to make sure I understand for my future reference: suppose there exists a filter construction corruption and detect_filter_construct_corruption=true catches it in db_stress, which now will lead to db_stress run terminates and core dumps. Will this test run be classified as "crash test failure" (like the one we need to investigate into)? If we do, do we need to change the way of handling such corruption happening in db_stress?

hx235 commented:
@pdillinger 
Update:
[[Design] Artifact release point.pdf](https://github.com/facebook/rocksdb/files/7918499/Design.Artifact.release.point.pdf)
1. Ready for review - IMO the most challenging part of this PR is to ensure each filter construction material and its cache reservation is released properly despite presence of corruption detection and the actual corruption. The "multiple-Finish-before-multiple-block checksum" implementation of Partitioned Filter also causes us some pain... Therefore I attached a summary of each artifact release point in all cases in my design, serving as a high-level design document. 
2. Left some notes and 1 question about stress test
3. ci/circleci: build-macos failure seems to be related to my https://github.com/facebook/rocksdb/pull/9393, which this PR is on top of. Need to look into that.
4. Haven't rebased on the latest https://github.com/facebook/rocksdb/pull/9393 as there is some discussion about touching a very hot write path. For this PR, I am good with building on (or not) top of  #9393.

Thanks!

hx235 commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -1231,92 +1231,106 @@ void BlockBasedTableBuilder::WriteRawBlock(const Slice& block_contents,
-        if (!s.ok()) {
-          r->SetStatus(s);
-        }
+
+  {
+    IOStatus io_s = r->file->Append(block_contents);
+    if (!io_s.ok()) {
+      r->SetIOStatus(io_s);
+      r->SetStatus(io_s);
+      return;
+    }
+  }
+
+  std::array<char, kBlockTrailerSize> trailer;
+  trailer[0] = type;
+  uint32_t checksum = ComputeBuiltinChecksumWithLastByte(
+      r->table_options.checksum, block_contents.data(), block_contents.size(),
+      /*last_byte*/ type);
+
+  if (block_type == BlockType::kFilter) {
+    Status s = r->filter_builder->MaybePostVerifyFilter(block_contents);
+    if (!s.ok()) {
+      r->SetStatus(s);
+      return;
+    }
+  }
```
Note: relevant to what you [mentioned](https://github.com/facebook/rocksdb/pull/9393#issuecomment-1015720627) about early-return and RAII, this is a good example of how early-return on append failure could have messed things up if we didn't have `ResetFilterBitsFilter`.

hx235 commented:
@pdillinger 
Update:
[[Design] Filter construction artifacts release point.pdf](https://github.com/facebook/rocksdb/files/7918617/Design.Filter.construction.artifacts.release.point.pdf)

- Ready for review - IMO the most challenging part of this PR is to ensure each filter construction material and its cache reservation is released properly as soon as the material becomes useless, under cases of turning on corruption detection and experiencing the actual corruption. The "multiple-Finish-before-multiple-block checksum" and "1 partition corruption = whole filter useless" implementation of Partitioned Filter didn't make our life easy. Therefore I attached a summary of each material release point in all cases in my design, serving as a high-level design document. 
- Left some notes and 1 question about stress test
- ci/circleci: build-macos failure might or might not relate to my https://github.com/facebook/rocksdb/pull/9393, which this PR is on top of. Haven't looked into that.
- Haven't rebased on the latest #9393 as there is some [discussion](https://github.com/facebook/rocksdb/pull/9393#pullrequestreview-857570603) about touching a very hot write path. For this PR, I am good with building on top of it or not.

Thanks!

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -283,11 +357,20 @@ class FastLocalBloomBitsBuilder : public XXPH3FilterBitsBuilder {
-    // Release cache for hash entries
-    hash_entry_cache_res_bucket_handles_.clear();
+    bool keep_entries_for_postverify = detect_filter_construct_corruption_;
+    if (!keep_entries_for_postverify) {
```
Self-documenting code 👍

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -203,9 +216,30 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
-  // A deque avoids unnecessary copying of already-saved values
-  // and has near-minimal peak memory use.
-  std::deque<uint64_t> hash_entries_;
+  Status MaybeVerifyHashEntriesChecksum() {
```
Ideally, the values would be verified as we are iterating over them in construction. That would be faster and leave fewer spaces for undetected memory/CPU corruption. But I understand it is difficult to inject into the modular Ribbon construction code. So I'm ok with just a comment (TODO?) pointing out the potential improvement, or make the improvement for Bloom only, where construction CPU efficiency is more critical.

Actually, `ResetAndFindSeedToSolve` is a template function that can use a custom iterator. If we have a custom iterator that tracks the xor checksum as it iterates, that might work for both Bloom and Ribbon.

hx235 replied:
Added to TODO - I will include Bloom as follow-up work too since changing it now changes the error injection approach in my test now.

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -381,9 +469,13 @@ class FastLocalBloomBitsBuilder : public XXPH3FilterBitsBuilder {
-      hash_entries_.pop_front();
+      uint64_t h = hash_entries_info_.entries.front();
+      hash_entries_info_.entries.pop_front();
+      if (keep_entries_for_postverify) {
```
This approach of pushing entries back on as we pop them off is not ideal for efficiency (or reliability) either. Conceptually, we are always iterating over the entries but sometimes destroying them as we go and sometimes keeping them. This could be a template function that takes different kind of iterators to use.

Actually, destroying them as we go may not be critical to efficiency (can be tested with filter_bench), so maybe template not needed.

hx235 replied:
Make sense - we can change deque to list structure as well as do not destroy entries in this function. Let's see how the performance change. Fixing it.

hx235 replied:
EDIT: please ignore my prev comment in here - it is incorrect.

pdillinger reviewed (changes requested):
Please also add to db_crashtest.py

Incomplete review, but some first round feedback.

hx235 commented on table/block_based/filter_policy.cc:
```diff
@@ -283,11 +357,20 @@ class FastLocalBloomBitsBuilder : public XXPH3FilterBitsBuilder {
+        if (status) {
+          *status = verify_hash_entries_checksum_status;
+        }
+        return FinishAlwaysFalse(buf);
```
Just want to double check, if we have `FinishAlwaysTrue` here and it is later passed to [BuiltinFilterPolicy::GetFilterBitsReader](https://github.com/facebook/rocksdb/blob/main/table/block_based/filter_policy.cc#L1315-L1321), will it still fall under the `return new AlwaysFalseFilter();` case?  

And also to double check, the content of the filter by itself without how we read it in `GetFilterBitsReader` does not dictate effect of "FinishAlwaysTrue" nor "FinishAlwaysTrue" - am I right?

hx235 replied:
Any update on this?

pdillinger replied:
A filter from FinishAlwaysTrue should reach this case: https://github.com/facebook/rocksdb/blob/a3de7ae49f59d517e3640695ca250d61e14feb56/table/block_based/filter_policy.cc#L1348

hx235 replied:
But we got something like this at top before that line of code:

```
  if (len_with_meta <= kMetadataLen) {
    // filter is empty or broken. Treat like zero keys added.
    return new AlwaysFalseFilter();
  }
```

pdillinger replied:
> ```
>   if (len_with_meta <= kMetadataLen) {
> ```

Oh, you caught me in an off-by-one error. Thanks

hx235 commented on table/block_based/filter_policy.cc:
```diff
@@ -259,9 +270,9 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
-    // and has near-minimal peak memory use.
-    std::deque<uint64_t> entries;
+    // A list avoids unnecessary copying of already-saved values
+    // and has less peak memory use than vector.
```
Question: 
Is the "near-minimal peak memory use." referring to the underlying impl of std::deque not pre-allocating memory slots as std::vector? I reworded for the list case.

hx235 replied:
Please ignore

hx235 commented:
Update: ready for review after signals are cleared
- Fixed xor_checksum naming, added to db_crashtest.py and not dropping hash entries in AddAllEntries (not dropping hash entries might affect performance VS not pushing it back might improve performance)
  - Update bloom related db bench result (observing overall 4-7% improvement compared with my last PR version, even some compared with the upstream/main. So not dropping hash entries along the way and drop them in bulk later might actually improve the performance 👍)
- Left a question for FinishAlwaysTrue
- Left some questions for future clarification on follow-up
- Rebase on latest https://github.com/facebook/rocksdb/pull/9393

hx235 commented:
Rebase on merged #9393

pdillinger commented on table/block_based/filter_policy.cc:
```diff
@@ -203,9 +220,37 @@ class XXPH3FilterBitsBuilder : public BuiltinFilterBitsBuilder {
+    } else {
+      // Since these hash entries are corrupted and they will not be used
+      // anymore, we can reset them and release memory.
+      ResetEntries();
```
I worry slightly that if someone ignores the status and proceeds with construction, we will get an "always false" (empty) filter rather than a safe "always true" filter. But I'm not sure it's worth an extra field and extra checks. Maybe just a stern warning in the comments right here. ;)

hx235 replied:
> Maybe just a stern warning in the comments right here
Not sure I understand why we need a comment in here instead of in https://github.com/facebook/rocksdb/pull/9342/files#diff-90d330d77a09b375bb5370051e4529ee67c905d7b72b2b599de722edab9a44f7R68-R71. I can add something like "A corrupted filter should not be further used since it's unsafe". 

> But I'm not sure it's worth an extra field and extra checks.
What are those extra checks for?

hx235 replied:
EDIT: I believe we will still get a safe "always true" filter due to  [this](https://github.com/facebook/rocksdb/pull/9342/files#diff-96639dd177bb4e9d0f353540edef15d418263a1ce94408b0b8d4cfe4711b7f1aR374). ResetEntries() is just for releasing artifacts while the returned filter should still be [AlwaysTrueFilter](https://github.com/facebook/rocksdb/pull/9342/files#diff-96639dd177bb4e9d0f353540edef15d418263a1ce94408b0b8d4cfe4711b7f1aR374). 

I don't think clearing hash entries affects the filter type - for example, we also clear hash entries before returning a normal filter when there is no corruption errr. However, I am still adding a warning to Finish API of "A corrupted filter should not be further used since it's unsafe" cuz it's good to have.

pdillinger commented on db/db_bloom_filter_test.cc:
```diff
@@ -1102,6 +1222,140 @@ TEST_P(DBFilterConstructionReserveMemoryTestWithParam, ReserveMemory) {
+  Options options = CurrentOptions();
+  // We set write_buffer_size big enough so that flush won't be triggered before
+  // we manually trigger it for clean testing
+  options.write_buffer_size = 640 << 20;
```
This is big for a running unit test, but maybe ok. Can we reduce it?

pdillinger commented on db/db_bloom_filter_test.cc:
```diff
@@ -1102,6 +1222,140 @@ TEST_P(DBFilterConstructionReserveMemoryTestWithParam, ReserveMemory) {
+            (std::deque<uint64_t>*)arg;
+        assert(!hash_entries_to_corrupt->empty());
+        *(hash_entries_to_corrupt->begin()) =
+            *(hash_entries_to_corrupt->begin()) ^ static_cast<uint64_t>(1);
```
Can write `uint64_t{1}`

pdillinger commented on db/db_bloom_filter_test.cc:
```diff
@@ -1102,6 +1222,140 @@ TEST_P(DBFilterConstructionReserveMemoryTestWithParam, ReserveMemory) {
+    ASSERT_OK(Put(Key(i), Key(i)));
+  }
+  SyncPoint::GetInstance()->SetCallBack(
+      "XXPH3FilterBitsBuilder::MaybeVerifyHashEntriesChecksum::PreVerification",
```
I would prefer to inject code into filter construction (e.g. code that is executed without `detect_filter_construct_corruption`) because that should be more realistic IMHO.

pdillinger commented on db/db_bloom_filter_test.cc:
```diff
@@ -1102,6 +1222,140 @@ TEST_P(DBFilterConstructionReserveMemoryTestWithParam, ReserveMemory) {
+        std::unique_ptr<FilterBitsReader>* bits_reader_p = arg_pair->first;
+        std::string filter_content = arg_pair->second;
+
+        corrupted_filter_content = "@#$" + filter_content;
```
This corruption re-arranges all the data, changing its size. I would prefer a more subtle corruption, such as clearing the first 8 bytes to zero. Not a big deal.

pdillinger reviewed (approved):
Overall looks good. A few nits and concerns. Thanks

hx235 commented on include/rocksdb/filter_policy.h:
```diff
@@ -101,6 +130,9 @@ class FilterBitsReader {
   }
+
+  // Check if the hash of the entry match the bits in filter
+  virtual bool HashMayMatch(const uint64_t /* h */) { return true; }
```
It's on the reader side but we should have a  BuiltinFilterBitsReader - fixing it.

hx235 replied:
Actually, I just remembered the reason for putting it under FilterBitsReader is we are calling HashMayMatch on the returned std::unique_ptr<FilterBitsReader> bits_reader(
      filter_policy_->GetFilterBitsReader(filter_content));.

pdillinger replied:
Actually, your description has reminded me that someone can create a wrapper FilterPolicy that returns a wrapper FilterBitsReader.

I suggest we make the body of BuiltinFilterPolicy::GetFilterBitsReader into a static (non-virtual) function for invoking here, and have BuiltinFilterPolicy::GetFilterBitsReader invoke that function.

hx235 commented on include/rocksdb/table.h:
```diff
@@ -365,6 +365,14 @@ struct BlockBasedTableOptions {
   // This must generally be true for gets to be efficient.
   bool whole_key_filtering = true;
 
+  // If true, detect corruption during (new) Bloom Filter and
```
Got you - just saw your #9363

hx235 replied:
> This is an extra check (as in `DisableExtraChecks`) 

The `filter_detect_corruption` is not currently under `DisableExtraChecks` 
https://github.com/facebook/rocksdb/pull/9363/ yet IIUC. Maybe we should remove "(as in DisableExtraChecks)" to avoid illusion that the feature was included. If we ever add this feature under  `DisableExtraChecks` due to performance concern, we list this in the comment of `DisableExtraChecks` as "this includes feature x,y,z and filter_detect_corruption".

pdillinger replied:
Good points.

We know this feature has a CPU cost, so I suggest adding it to `DisableExtraChecks` even though this feature is currently disabled by default. Or you can say "(like those in DisableExtraChecks)" or remove the reference to DisableExtraChecks. Any of those is OK.

hx235 replied:
Will remove the reference add it to DisableExtraChecks in another quick PR.

hx235 commented:
Update: 
- Fixed all nits/concern except for returning AlwaysTrueFilter VS AlwaysFalseFilter on corruption return (left question for that). Will fix the comment about empty filter along with that after getting the answer.
- Rebased

hx235 commented:
Update:
- Refactored out and called the static BuiltinFilterPolicy::GetBuiltinFilterBitsReader
  - No need to store FilterPolicy in FilterBitsBuilder anymore
- Updated AlwaysTrueFilter length to be 6
- Updated BlockBasedTableOptions::detect_filter_corruption API related to DisableExtraCheck
- Added a warning for not using a filter constructed with corruption 
- Merge after signals are clear
